hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/serialization/DelegatingSerializationService.java

```diff
@@ -108,33 +108,57 @@ public PortableContext getPortableContext() {
 
     @Override
     public SerializerAdapter serializerFor(Object object) {
-        Class<?> clazz = object.getClass();
-        SerializerAdapter serializer = serializersByClass.get(clazz);
+        Class<?> clazz = object == null ? null : object.getClass();
+
+        SerializerAdapter serializer = null;
+        if (clazz != null) {
+            serializer = serializersByClass.get(clazz);
+        }
         if (serializer == null) {
-            serializer = delegate.serializerFor(object);
-            if (serializer == null) {
-                throw active ?
-                        new HazelcastSerializationException("There is no suitable serializer for " + clazz) :
-                        new HazelcastInstanceNotActiveException();
+            try {
+                serializer = delegate.serializerFor(object);
+            } catch (HazelcastSerializationException hse) {
+                throw handle(clazz, hse);
             }
         }
+        if (serializer == null) {
+            throw active ? new MissingSerializer(clazz) : new HazelcastInstanceNotActiveException();
+        }
         return serializer;
     }
 
+    private RuntimeException handle(Class<?> clazz, HazelcastSerializationException e) {
+        String message = e.getMessage();
+        if (message != null && message.startsWith("There is no suitable serializer for")) {
+            return new MissingSerializer(clazz);
+        }
+        return e;
+    }
+
     @Override
     public SerializerAdapter serializerFor(int typeId) {
         SerializerAdapter serializer = serializersById.get(typeId);
         if (serializer == null) {
-            serializer = delegate.serializerFor(typeId);
-            if (serializer == null) {
-                throw active ?
-                        newHazelcastSerializationException(typeId) :
-                        new HazelcastInstanceNotActiveException();
+            try {
+                serializer = delegate.serializerFor(typeId);
+            } catch (HazelcastSerializationException hse) {
+                throw handle(typeId, hse);
```

Comment:
We should not depend on the exact text from the superclass. We also lose the stack trace of the original exc.

I'd rather just wrap the caught exception, use the original message and append a text to it.
```suggestion
                throw new HazelcastSerializationException(hse.getMessage() + ". You can register a job-specific serializer using JobConfig.registerSerializer()", hse);
```

If we also dispose the delegate (which we should), we wouldn't have to handle the inactive case: the delegate will throw `HazelcastInstanceNotActiveException` directly.

Reply:
We should not `dispose()` delegate as it's a _global_ serializer. I have addressed the remaining comment.